File: src/utils/commandParser.ts
import { ParsedCommand } from '../types';

export function parseCommand(input: string): ParsedCommand {
  const lower = input.toLowerCase().trim();

  // Help command
  if (
    lower === 'help' ||
    lower === 'bantuan' ||
    lower === 'tolong' ||
    lower.includes('apa yang bisa') ||
    lower.includes('fitur') ||
    lower.includes('perintah')
  ) {
    return { action: 'help', params: {} };
  }

  // Create wallet
  if (
    lower.includes('buat wallet') ||
    lower.includes('create wallet') ||
    lower.includes('generate wallet') ||
    lower.includes('wallet baru') ||
    lower.includes('bikin wallet') ||
    lower.includes('new wallet') ||
    lower.includes('buat dompet') ||
    lower.includes('dompet baru')
  ) {
    return { action: 'create_wallet', params: {} };
  }

  // Import wallet
  if (
    lower.includes('import wallet') ||
    lower.includes('import key') ||
    lower.includes('import private') ||
    lower.includes('impor wallet') ||
    lower.includes('masukkan private key') ||
    lower.includes('pakai private key')
  ) {
    // Try to extract private key from input
    const hexMatch = input.match(/0x[a-fA-F0-9]{64}/);
    const rawHexMatch = input.match(/[a-fA-F0-9]{64}/);
    const privateKey = hexMatch ? hexMatch[0] : rawHexMatch ? rawHexMatch[0] : '';

    return {
      action: 'import_wallet',
      params: { privateKey },
    };
  }

  // Check balance
  if (
    lower.includes('cek saldo') ||
    lower.includes('check balance') ||
    lower.includes('saldo') ||
    lower.includes('balance') ||
    lower.includes('berapa saldo') ||
    lower.includes('lihat saldo')
  ) {
    return { action: 'balance', params: {} };
  }

  // Send BNB
  if (
    lower.includes('kirim') ||
    lower.includes('send') ||
    lower.includes('transfer') ||
    lower.includes('bayar') ||
    lower.includes('pay')
  ) {
    // Extract amount
    const amountMatch = input.match(/(\d+\.?\d*)\s*(?:bnb|BNB)?/);
    const amount = amountMatch ? amountMatch[1] : '';

    // Extract address
    const addressMatch = input.match(/0x[a-fA-F0-9]{40}/);
    const toAddress = addressMatch ? addressMatch[0] : '';

    return {
      action: 'send',
      params: { amount, toAddress },
    };
  }

  return { action: 'unknown', params: {} };
}

export function getHelpMessage(): string {
  return `🤖 **VibeAgent - Perintah yang Tersedia:**

💰 **Wallet:**
• "Buat wallet baru" - Generate wallet baru
• "Import private key 0x..." - Import wallet dari private key
• "Cek saldo" - Lihat saldo BNB

💸 **Transaksi:**
• "Kirim 0.01 BNB ke 0x..." - Transfer BNB
• "Transfer 0.5 BNB ke 0x..." - Transfer BNB

📌 **Catatan:**
• Semua transaksi berjalan di BSC Testnet
• Dapatkan BNB testnet di: https://testnet.bnbchain.org/faucet-smart
• Ketik "help" untuk melihat perintah ini lagi`;
}

File: VibeAgent-Android/app/src/main/java/com/vibeagent/app/util/CommandParser.kt
package com.vibeagent.app.util

data class ParsedCommand(
    val action: String,
    val params: Map<String, String> = emptyMap()
)

class CommandParser {

    fun parse(input: String): ParsedCommand {
        val lowerInput = input.lowercase().trim()

        return when {
            // Help command
            lowerInput == "help" || lowerInput == "bantuan" || lowerInput == "?" -> {
                ParsedCommand("help")
            }

            // Create wallet (English + Indonesian aliases)
            lowerInput.contains("create wallet") ||
            lowerInput.contains("create new wallet") ||
            lowerInput.contains("new wallet") ||
            lowerInput.contains("buat wallet") || 
            lowerInput.contains("wallet baru") -> {
                ParsedCommand("create_wallet")
            }

            // Import wallet
            lowerInput.contains("import") && lowerInput.contains("private key") -> {
                val privateKey = extractPrivateKey(input)
                ParsedCommand("import_wallet", mapOf("privateKey" to privateKey))
            }

            // Check balance (English + Indonesian aliases)
            lowerInput.contains("check balance") ||
            lowerInput.contains("balance") ||
            lowerInput.contains("cek saldo") ||
            lowerInput.contains("saldo") -> {
                ParsedCommand("balance")
            }

            // Send BNB (English + Indonesian aliases)
            lowerInput.contains("send") || lowerInput.contains("transfer") || lowerInput.contains("kirim") -> {
                val (amount, address) = extractSendParams(input)
                ParsedCommand("send", mapOf("amount" to amount, "toAddress" to address))
            }

            else -> ParsedCommand("unknown")
        }
    }

    private fun extractPrivateKey(input: String): String {
        // Look for hex string starting with 0x
        val hexPattern = Regex("0x[a-fA-F0-9]{64}")
        val match = hexPattern.find(input)
        if (match != null) {
            return match.value
        }

        // Look for 64-character hex string without 0x prefix
        val plainHexPattern = Regex("[a-fA-F0-9]{64}")
        val plainMatch = plainHexPattern.find(input)
        return if (plainMatch != null) {
            "0x${plainMatch.value}"
        } else {
            ""
        }
    }

    private fun extractSendParams(input: String): Pair<String, String> {
        // Extract amount (number followed by BNB)
        val amountPattern = Regex("([0-9]+\\.?[0-9]*)\\s*(?:bnb)?", RegexOption.IGNORE_CASE)
        val amountMatch = amountPattern.find(input)
        val amount = amountMatch?.groupValues?.get(1) ?: ""

        // Extract address (0x followed by 40 hex characters)
        val addressPattern = Regex("0x[a-fA-F0-9]{40}")
        val addressMatch = addressPattern.find(input)
        val address = addressMatch?.value ?: ""

        return Pair(amount, address)
    }
}
